### DISCUSSION TITLE
Understanding adapter types and issues with codelldb

### DISCUSSION THREAD

**Discussion by asmodeus812** [mfussenegger/nvim-dap#626] [Q&A]
Hi, Your help will be greatly appreciated, I am trying to figure out why i am facing issues attaching/starting codelldb and other similar adapters which implement the protocol the normal regular way, and have to resort to spawning the process and then attaching !

This below, does not seem to work, as i understand it from the source what it does it spawn the process and then tries to initialize the session by sending request to the server, but it fails with server did not respond or was too slow or something similar
```Lua
adapter = {
        id = "codelldb",
        type = "executable",
        enrich_config = enricher,
        command = 'path-to-codelldb-exe',
        args = { "--port", "13000" }
    },
```

This however works, and the presumption here is it should be doing something pretty similar. Start the executable and then connect over to it or attach to it.
```Lua
        adapter = {
            type = "server",
            host = "127.0.0.1",
            port = "${port}",
            enrich_config = enricher,
            executable = {
                command = 'path-to-codelldb-exe',
                args = { "--port", "${port}" }
            }
        },
```

Now since, codelldb does support the protocol why is it not possible to just simply use the inital approach and why is the second one working, what are the differences flow wise ?

Another question is the "id" field at the adapter level, some adapters i haven't been able to make work without that id field i don't know why, can you explain when and where is the id field necessary and why ? One such adapter is the vsctools OpenDebugAD7. 

```Lua
adapter = {
        id = "cppdbg",
        type = "executable",
        enrich_config = enricher,
        options = { detached = false },
        command = "path-to-OpenDebugAD7"
    },
    configurations = {
        default_launch = {
            type = "cppdbg",
            request = "launch",
            stopOnEntry = false,
            cwd = "${workspaceFolder}",
            name = "Default launch (vscpptools)",
            program = function()
                // input the program executable to debug
            end,
            MIMode = "gdb",
            miDebuggerPath = "path-to-gdb-executable",
            setupCommands = {
                {
                    description = 'enable pretty printing',
                    text = '-enable-pretty-printing',
                    ignoreFailures = true
                },
            }
        },
    },
```

**Comment by mfussenegger** [comment]
This is somewhat explained in the docs:

> The `Adapter` needs to contain a `type`, which can be one of:
> 
> - `executable`, to indicate that nvim-dap must launch the debug adapter. In
>   this case nvim-dap will spawn the given process and communicate with it using
>   stdio.
> 
> - `server`, to indicate that nvim-dap can connect to an already-running
>   debug adapter via TCP.

And from https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(via--codelldb)#adapter-definition:

> codelldb uses TCP for the DAP communication - that requires using the server type for the adapter definition


`execute` spawns the adapter and uses STDIO as transport
`server` uses TCP as transport, and it can optionally spawn the adapter.

codelldb doesn't support using STDIO for the DAP communication.


> Another question is the "id" field at the adapter level, some adapters i haven't been able to make work without that id field i don't know why, can you explain when and where is the id field necessary and why 


It's used for the `adapterID` in the initialize request: https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Initialize

I don't know why some adapters require a certain value. You'd have to check their implementation or ask their maintainers.

**Reply by asmodeus812** [reply]
Okay, i see, so some debug server's implementing dap are only communicating over TCP and there is no way to establish the connection using the usual type='executable' as for example some other (mostly the js based adapters) work, so we use type='server' with executable (dap will manage the executable's lifetime) or type='server' no executable (used to attach to an already running server). 
On another note is it safe to pass id field to all adapters to be safe, or is that going to break some servers ? 

EDIT: To clarify, the server + executable adapter config only exists to allow TCP communication because some servers don't / can't use the regular executable type adapter config approach, is that correct ?
